components/lib/push-firebase/src/main/java/mozilla/components/lib/push/firebase/AbstractFirebasePushService.kt

```diff
@@ -0,0 +1,61 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ *  License, v. 2.0. If a copy of the MPL was not distributed with this
+ *  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package mozilla.components.lib.push.firebase
+
+import android.content.Context
+import com.google.android.gms.common.ConnectionResult
+import com.google.android.gms.common.GoogleApiAvailability
+import com.google.firebase.iid.FirebaseInstanceId
+import com.google.firebase.messaging.FirebaseMessaging
+import com.google.firebase.messaging.FirebaseMessagingService
+import com.google.firebase.messaging.RemoteMessage
+import mozilla.components.concept.push.EncryptedPushMessage
+import mozilla.components.concept.push.PushProvider
+import mozilla.components.concept.push.PushService
+
+abstract class AbstractFirebasePushService : FirebaseMessagingService(), PushService {
+
+    init {
+        start() // Allow the app to choose when to start? This is harder to allow than it looks..
+    }
+
+    final override fun start() {
+        FirebaseMessaging.getInstance().isAutoInitEnabled = true
+    }
+
+    override fun onNewToken(newToken: String) {
+        PushProvider.requireInstance.onNewToken(newToken)
+    }
+
+    override fun onMessageReceived(remoteMessage: RemoteMessage?) {
+        remoteMessage?.let {
+            val message = EncryptedPushMessage.invoke(
+                it.data.getValue("chid"),
+                it.data.getValue("body"),
+                it.data.getValue("con"),
+                it.data["enc"],
+                it.data["cryptokey"]
+            )
+            PushProvider.requireInstance.onMessageReceived(message)
+        }
+    }
+
+    final override fun stop() {
+        FirebaseMessaging.getInstance().isAutoInitEnabled = false
+    }
+
+    final override fun forceRegistrationRenewal() {
```

Comment:
It's not very clear to me how this works. Can you add a detailed comment which explains underlying behaviour?

From the docs: "when auto init is enabled, [.... fcm] generates a new token when it is deleted (which prevents deleteInstanceId() from stopping the periodic sending of data)".

So you stop this process first, then call `deleteInstanceId`. From the docs:

"Delete the Instance ID and the data associated with it.

This stops the periodic sending of data to the Firebase backend started when the Instance ID was generated, unless another library that requires InstanceId (like FCM, RemoteConfig or Analytics) is used or it's configured to be executed automatically.

A new Instance ID is generated asynchronously if Firebase Cloud Messaging auto-init is enabled."

So, second paragraph implies that this _may_ be a no-op operation? And last sentence implies that you don't need to switch `isAutoInitEnabled` flag at all! You could have just called a delete method by itself, and it would regenerate stuff (same as re-enabling auto-init after deleting).

That's my interpretation of the docs, I haven't actually used this API. Can you explain how you think it works?